### PR TITLE
LineIntersector: Fixed invalid read on mixed-dimensionality inputs

### PR DESCRIPTION
**Pull request by dbaston**
Fixes https://github.com/libgeos/geos/issues/1074

### CODE REVIEW

File: include/geos/algorithm/LineIntersector.h [hunk 1]
```diff
@@ -589,8 +589,20 @@ class GEOS_DLL LineIntersector {
...
-            ptInt = static_cast<const C1&>(nearestEndpoint(p1, p2, q1, q2));
+            const geom::CoordinateXY& nearest = nearestEndpoint(p1, p2, q1, q2);
+#if __cplusplus >= 201703L
+            if constexpr (std::is_same<C1, C2>::value) {
+#else
+            if (std::is_same<C1, C2>::value) {
+#endif
+                ptInt = static_cast<const C1&>(nearest);
+            } else {
+                if (&ptInt == static_cast<const geom::CoordinateXY*>(&p1) || &ptInt == static_cast<const geom::CoordinateXY*>(&p2)) {
```
Comment (rouault) [inline include/geos/algorithm/LineIntersector.h:600]: Probably due to my lack of familiarity with that code, but I can't make sense of those address comparisons.  ptInt is a new object created at line 590. How could its address be the same as p1 or p2 ?

Comment (dbaston) [reply include/geos/algorithm/LineIntersector.h:600]: It couldn't! The wrong variable is being compared here.

Comment (dbaston) [inline include/geos/algorithm/LineIntersector.h:600]:
```suggestion
                if (&nearest == static_cast<const geom::CoordinateXY*>(&p1) || &nearest == static_cast<const geom::CoordinateXY*>(&p2)) {
```

File: tests/unit/capi/GEOSIntersectionTest.cpp [hunk 2]
```diff
@@ -157,5 +157,15 @@ void object::test<8>()
...
 }
 
+template<>
+template<>
+void object::test<9>()
+{
+    geom1_ = fromWKT("GEOMETRYCOLLECTION(POINT(0 0 59.083333333333336), POINT(1 1 59.083333333333336), LINESTRING(2 4 111 1, 2 223 22 2), POLYGON((42 -2.222222223222222e+155 111 1, 2 3 1 1, 2 4 111 NaN, 42 -2.222222223222222e+155 111 NaN)), POLYGON((2.2222222212222224e+149 4 111 1, 42 -2.222222222222222e+149 22 2, 2 4 111 1, 22 2222222222222223 22 NaN, 2.2222222212222224e+149 4 111 NaN)))");
+    geom2_ = fromWKT("LINESTRING (1.6409301755752343e+149 -5.298190649085575e+148, 1.6666666660752404e+149 -5.55555555396982e+148)");
+
+    result_ = GEOSIntersection(geom1_, geom2_);
```
[the remark this answers is not in the document]
Comment (dbaston) [reply tests/unit/capi/GEOSIntersectionTest.cpp:170]: The purpose of the test is to verify that no crash occurs - I'm not sure the specific result is important. But I added asserts to verify that the result is non-null and has the correct dimensionality.

**Pull request merged**